### ISSUE TITLE
Rector analysis fails when using phar-distributed phpunit

### ISSUE THREAD

**Issue by rela589n** [rectorphp/rector#8435]
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.19.2             |

I have installed phpunit as a [phar archive](https://docs.phpunit.de/en/10.5/installation.html#php-archive-phar), and configured `autoload_dev.php` script that just requires a phar-file: `require_once __DIR__.'/../bin/phpunit.phar';` to integrate phpunit.phar with phpstan and psalm static analysis tools. The same result may be achieved with `bootstrapFiles` phpstan configuration and `autoload` psalm attribute where `phpunit.phar` location may be specified so that tools will be aware of classes and methods provided by phpunit, but I created `autoload_dev.php` to keep things unified.

Finally I stumbled upon following issue when running rector with `autoload_dev.php` enabled:

```shell
> vendor/bin/rector process --dry-run --clear-cache -vvv
Script vendor/bin/rector process --dry-run --clear-cache -vvv handling the ci:rector event returned with error code 2
```

Surely, It disappears if I remove `require_once` statement from `tests/autoload_dev.php`:

```diff
-require_once __DIR__.'/../bin/phpunit.phar';
```

But then I get following diff from rector (RemoveParentCallWithoutParentRector):

```diff
    ---------- begin diff ----------
@@ @@

     private StackMiddleware $stack;

-    protected function setUp(): void
+    private function setUp(): void
     {
-        parent::setUp();
-
         $container = self::getContainer();

         /** @var ExceptionalValidationMiddleware $middleware */
    ----------- end diff -----------

Applied rules:
 * RemoveParentCallWithoutParentRector
 * PrivatizeFinalClassMethodRector
```

Basically It removes everything related to phpunit, since it is not aware of methods.

## Minimal PHP Code Causing Issue

The example is not that minimal, but you may reproduce the issue on that commit:

https://github.com/phphd/exceptional-validation-bundle/commit/e9e5e4e70728edaec40858fdba31529e550c6000

Just run `composer update` and `composer ci:rector` and see what happens.

## Expected Behaviour

It is expected that rector would handle phar-loaded classes the same way as phpstan and psalm does.

**Comment by TomasVotruba** [comment]
Thanks for reporting. Agreed, that it should be working with phar as well, if the file is provided to Rector bootstrap autoload :+1: 

> Script vendor/bin/rector process --dry-run --clear-cache -vvv handling the ci:rector event returned with error code 2

This error code is expected, as it suggest code has changed. Error code 1 is an error. What is the provided diff?

<br>

Could you share a reproducible repository? There is a lot going on, so we know the exact code that is being run.

**Comment by samsonasik** [comment]
@TomasVotruba the repo is here https://github.com/phphd/exceptional-validation-bundle per issue description.

**Comment by TomasVotruba** [comment]
@samsonasik That is whole bundle with other dependencies that would have to be removed to isolate the issue. We look for minimal reproducer like PHPUnit phar + single class + single rule.

**Comment by rela589n** [comment]
Hi @TomasVotruba , 

Thanks for your prompt response!

I've just created a reproduction repository you may check out: https://github.com/rela589n/rector-8435

**Comment by samsonasik** [comment]
I created PR to allow directly load phar on bootstrap files config

- https://github.com/rectorphp/rector-src/pull/5518

**Comment by rela589n** [comment]
Hi @samsonasik 

Thank you for your time!

I've just checked out `fix-load-phar` brach and unfortunately it doesn't seem to fix the problem. Definitely, now rector prints the output, but phar doesn't look like having been loaded correctly, since current output is:

```diff
-    protected function setUp(): void
+    private function setUp(): void
```

![image](https://github.com/rectorphp/rector/assets/41589422/71dd270b-3113-47fa-98c9-1692692ad235)

I appreciate any ideas you might have on this matter.

Thank you and have a nice day!

**Comment by samsonasik** [comment]
Latest rector allow to load bootstrap by:

```php
    $rectorConfig->bootstrapFiles([
        __DIR__ . '/bin/phpunit.phar',
    ]);
```

see PR:

- https://github.com/rectorphp/rector-src/pull/5518

Ensure that you remove unused autoad file:

```diff
-        },
-        "files": [
-            "tests/autoload_dev.php"
-        ]
```

Then, remove the file 

```
rm -rf tests/autoload_dev.php
```

and run composer update to get latest `"rector/rector": "^0.19.5":

```
composer update rector/rector:^0.19.5
```

**Comment by samsonasik** [comment]
Ok, I re-checked it, it may be due to rector provide the stubs file, which overlap it, 

https://github.com/rectorphp/rector/blob/c8283050410f039e061d99f8e2c311c677935feb/stubs-rector/PHPUnit/Framework/TestCase.php#L10

which should not, as it `Phar::loadPhar()` called early, re-open for more investigating.

**Comment by samsonasik** [comment]
Ok, the phar file seems need to be opened, to include all the files early before late detected by composer autoload , I will check how to get the files from the doc when i back to pc
https://www.php.net/manual/en/phar.using.intro.php

**Comment by samsonasik** [comment]
Ok, the reason why it always loaded is because you define 

```php
$rectorConfig->paths([__DIR__.'/']);
```

that will make it always loaded, even rector stubs is not loaded by bootstrap. I will create a PR to both rector-src and your repo for the solution :)

**Comment by samsonasik** [comment]
@rela589n I created new PR for real solution based on latest rector/rector:dev-main to your repo:

- https://github.com/rela589n/rector-8435/pull/2

**Comment by samsonasik** [comment]
Ok, re-open, `Phar::loadPhar()` actually doesn't make it included. Rector run and skipped by using latest dev-main because now parent never detected, so it looks like OK, while actually not.

**Comment by samsonasik** [comment]
I found that load phar check is not really needed, checkign wtih:

```php
            // load phar file
            if (\substr_compare($bootstrapFile, '.phar', -\strlen('.phar')) === 0) {
                require($bootstrapFile);
                var_dump(class_exists('PHPUnit\Framework\TestCase'));
```

got:

```
bool(true)
```

so just define directly on boostrap, I will create new PR for it.
value.

**Comment by samsonasik** [comment]
Hm.., not sure, it loaded `true`, but nothing to display:

```
➜  rector-8435 git:(real-fix) ✗ vendor/bin/rector --dry-run --clear-cache
➜  rector-8435 git:(real-fix) ✗ 
```

which should show either **"Rector is done!"** or **" [OK] 1 file would have changed (dry-run) by Rector "**

**Comment by samsonasik** [comment]
Trying include phar with `phar://` prefix got error:

```
➜  rector-8435 git:(real-fix) ✗ vendor/bin/rector --dry-run --clear-cache
PHP Warning:  Constant  already defined in /Users/samsonasik/www/rector-8435/bin/phpunit.phar on line 3003
Warning: Constant  already defined in /Users/samsonasik/www/rector-8435/bin/phpunit.phar on line 3003
PHP Warning:  Constant __PHPUNIT_PHAR__ already defined in /Users/samsonasik/www/rector-8435/bin/phpunit.phar on line 76
Warning: Constant __PHPUNIT_PHAR__ already defined in /Users/samsonasik/www/rector-8435/bin/phpunit.phar on line 76
PHP Warning:  Constant __PHPUNIT_PHAR_ROOT__ already defined in /Users/samsonasik/www/rector-8435/bin/phpunit.phar on line 77
Warning: Constant __PHPUNIT_PHAR_ROOT__ already defined in /Users/samsonasik/www/rector-8435/bin/phpunit.phar on line 77
```

**Comment by samsonasik** [comment]
I temporary revert the Phar load support on `BootstrapFilesIncluder` as it not really working well

- https://github.com/rectorphp/rector-src/pull/5529

Not sure how it should works.

Actually, include phpunit.phar (even require_once) already make class exists, but somehow, probably have constant conflict on the fly on definition that make it doesn't show anything.

I suggest to not use `phar` file, use normal `phpunit` from vendor so Reflection will always catch up :)

**Comment by rela589n** [comment]
@samsonasik , first of all, thanks for your effort. 

If the phar-support solution is not obvious now, maybe let's not waste the time finding it. 
I guess I'll use direct dev-dependency on phpunit in https://github.com/phphd/exceptional-validation-bundle/ then

If at some point in future rector would elaborate direct support for .phar, I'd expect it working both via `bootstrapFiles()` configuration as well as manual `require_once` statement as in the reproduction repository. The key point in having `autoload_dev.php` bootstrap file is that we don't have to define a separate configuration options for phpstan, psalm, rector, etc. in order to include `phpunit.phar` file, but have it in a single place instead.

Feel free to close this issue for now.

**Comment by samsonasik** [comment]
Thank you for understanding, there is no standard on phar file what file inside it that should be included, that make it possibly halted during require directly, in contrast with phpstan.phar, that has bootstrap.php inside, in the meantime, you can require-dev phpunit/phpunit itself so it always correctly autoloaded.

**Comment by TomasVotruba** [comment]
It seems the phar was loaded correctly, but something was included twice.

I've cleaned up your repository with following commit:
https://github.com/rela589n/rector-8435/pull/3

Now Rector and PHPStan pass well :)

**Comment by samsonasik** [comment]
@TomasVotruba that's not working, the `tests` needs in the `paths()`, not only `src`, and `parent::setUp` got removed:

```diff
1) tests/FooTest.php:10

    ---------- begin diff ----------
@@ @@
 {
     private Foo $foo;

-    protected function setUp(): void
+    private function setUp(): void
     {
-        parent::setUp();
-
         $this->foo = new Foo(234);
```

On bootstrap, the `.phar` by directly is loaded, but internally, it may has halt compiler which stop the process.

**Comment by TomasVotruba** [comment]
I see :+1:  I'll look again

**Comment by samsonasik** [comment]
FYI, When we use `Phar::loadPhar()`, it only load the manifest, not "execute" the phar.

Loop phar files won't work, as we don't know which files should be loaded first, and will end with `class not found` because of sorting.

To test, you can do:

```php
require $bootstrapFile;
class_exists('PHPUnit\Framework\TestCase`
```

that actually make `true` result, but process is halted.

In Phar, there is "->getStub()" which content of stubs that have `spl_autoload`, but that will make crazy regex check, and probably not all phar has that.

That's why it is not possible, as far as I tried.

**Comment by TomasVotruba** [comment]
So indeed, the phpunit phar is loaded and Rector runs. For some reason there is no output.
I assume PHPUnit somehow overrides Rector output :thinking:

**Comment by TomasVotruba** [comment]
@rela589n After few hours of debugging, not sure if we can find solution now.

I'd recommend switching to normal PHPUnit package as obvious solution. There is no reason not to use those 3 tools together as composer dependencies, as Rector and PHPStan require PHP 7.2+ and have no blockers.

**Comment by samsonasik** [comment]
`phpunit.phar` is used as executable, that's why it doesn't have file "bootstrap.php" or something2 autoload that user can use is as autoload

in contrast with **phpstan.phar** has "bootstrap.php" which people can use.
